### PR TITLE
Causal Conv1d

### PR DESCRIPTION
**Pull request by Haichao-Zhang**
This PR implements CausalConv1D.
This might be useful for standard autoregressive scenarios where RNN is typically used.

### CODE REVIEW

File: alf/layers.py [hunk 1]
```diff
@@ -705,6 +706,118 @@ def bias(self):
         return self._bias
 
 
+@alf.configurable
+class CausalConv1D(nn.Module):
```
Comment (hnyu) [inline alf/layers.py:710]: it seems that this is only dilated conv1d? why do we need to create a new layer class named as "CausalConv1D"? Would be better to explicitly explain the difference between a dilated conv1d and this class in the docstring.

Comment (Haichao-Zhang) [reply alf/layers.py:710]: > it seems that this is only dilated conv1d? why do we need to create a new layer class named as "CausalConv1D"? Would be better to explicitly explain the difference between a dilated conv1d and this class in the docstring.

Not really. dilation is just an optional choice. Even without dilation, it is still different from conv1d, in that it is causal. Will add more explanations in comments.

Comment (Haichao-Zhang) [reply alf/layers.py:710]: Now added more explanations.

File: alf/layers.py [hunk 2]
```diff
@@ -705,6 +706,137 @@ def bias(self):
...
+        initialization. An auto gain calculation might depend on the activation
+        following the causal conv1d layer.
+
+        Note that the main difference of causal conv v.s. standard conv is that
+        each temporal element in the convolutional output is causal w.r.t.
+        the temporal elements from input. For example, for a length L sequence
+        x with the shape of [B, C, L], and y = causal_conv(x), where the shape
+        of y is [B, C', L], by causal we mean y[..., l] only depends on
+        X[..., :l] (i.e. the past), and there is no dependency on X[..., :l]
+        (i.e. future) as in the standard non-causal convolution.
```
Comment (hnyu) [inline alf/layers.py]: Use correct formatting for these math notations.

File: alf/layers.py [hunk 3]
```diff
@@ -729,13 +730,27 @@ def __init__(self,
...
+        shift the input to the right (future) according to kernel size.
+
         Args:
             in_channels (int): channels of the input
             out_channels (int): channels of the output
             kernel_size (int): size of the kernel
             dilation (int): controls the spacing between the kernel points.
                 Please refer to here for a visual illustration:
                 https://github.com/vdumoulin/conv_arithmetic/blob/master/README.md
+            hide_current (bool): whether to hide the current by shifting the
```
Comment (hnyu) [inline alf/layers.py:751]: If the current info in the first layer is missing, how will the later layers get it?

File: alf/layers.py [hunk 4]
```diff
@@ -705,6 +706,137 @@ def bias(self):
...
+        shift the input to the right (future) according to kernel size.
+
+        Args:
+            in_channels (int): channels of the input
+            out_channels (int): channels of the output
+            kernel_size (int): size of the kernel
+            dilation (int): controls the spacing between the kernel points.
+                Please refer to here for a visual illustration:
+                https://github.com/vdumoulin/conv_arithmetic/blob/master/README.md
+            hide_current (bool): whether to hide the current by shifting the
```
Comment (hnyu) [inline alf/layers.py:753]: The need of hide_current seems better addressed by manipulating the input before feeding it to this layer?

File: alf/layers.py (hunk 2 above)
Comment (Haichao-Zhang) [reply alf/layers.py]: Fixed

File: alf/layers.py (hunk 4 above)
Comment (Haichao-Zhang) [reply alf/layers.py:753]: We can choose either way to do this. Here we just provide an option to the user.

File: alf/layers.py [hunk 5]
```diff
@@ -732,11 +732,13 @@ def __init__(self,
...
-        the temporal elements from input. For example, for a length L sequence
-        x with the shape of [B, C, L], and y = causal_conv(x), where the shape
-        of y is [B, C', L], by causal we mean y[..., l] only depends on
-        X[..., :l] (i.e. the past), and there is no dependency on X[..., :l]
-        (i.e. future) as in the standard non-causal convolution.
+        the temporal elements from input. For example, for a length :math:`L`
+        sequence :math:`x` with the shape of :math:`[B, C, L]`, and
+        :math:`y = causal_conv(x)`, where the shape of :math:`y` is
+        :math:`[B, C', L]`, by causal we mean :math:`y[..., l]` only depends on
+        :math:`X[..., :l]` (i.e. the past), and there is no dependency on
```
Comment (hnyu) [inline alf/layers.py]: this is actually python syntax, not latex syntax

Comment (Haichao-Zhang) [reply alf/layers.py]: changed

File: alf/layers.py [hunk 6]
```diff
@@ -705,6 +706,139 @@ def bias(self):
...
+            self._bn = nn.BatchNorm1d(out_channels)
+        else:
+            self._bn = None
+
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        """Initialize the parameters."""
+        if self._kernel_initializer is None:
+            variance_scaling_init(
```
Comment (hnyu) [inline alf/layers.py:804]: One last minor potential issue: variance_scaling_init calculates the gain by assuming the input is N(0,1). Now since we know we are always padding the input with zeros, this auto calculation might be problematic.

**Review by hnyu — APPROVED** [review]
